spec.html

```diff
@@ -24817,55 +25545,89 @@ <h1>
-        1. If _classBinding_ is not *undefined*, then
-          1. Perform ! _classEnv_.InitializeBinding(_classBinding_, _F_).
-        1. Set _F_.[[PrivateMethods]] to _instancePrivateMethods_.
-        1. Set _F_.[[Fields]] to _instanceFields_.
+        1. Let _instanceMetadataList_ be a new empty List.
+        1. Let _staticMetadataList_ be a new empty List.
+        1. Let _instanceExtraInitializers_ be a new empty List.
+        1. Let _staticExtraInitializers_ be a new empty List.
+        1. For each element _e_ of _staticElements_, do
+          1. If _e_.[[Kind]] is not ~field~, then
+            1. Let _result_ be Completion(ApplyDecoratorsAndDefineMethod(_F_, _e_, _staticMetadataList_, _staticExtraInitializers_, *true*)).
+            1. If _result_ is an abrupt completion, then
+              1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+              1. Return ? _result_.
+        1. For each element _e_ of _instanceElements_, do
+          1. If _e_.[[Kind]] is not ~field~, then
+            1. Let _result_ be Completion(ApplyDecoratorsAndDefineMethod(_proto_, _e_, _instanceMetadataList_, _instanceExtraInitializers_, *true*)).
+            1. If _result_ is an abrupt completion, then
+              1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+              1. Return ? _result_.
+        1. For each element _e_ of _staticElements_, do
+          1. If _e_.[[Kind]] is ~field~, then
+            1. Let _result_ be Completion(ApplyDecoratorsToElementDefinition(_F_, _e_, _staticMetadataList_, _staticExtraInitializers_, *true*)).
+            1. If _result_ is an abrupt completion, then
+              1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+              1. Return ? _result_.
+        1. For each element _e_ of _instanceElements_, do
+          1. If _e_.[[Kind]] is ~field~, then
+            1. Let _result_ Completion(ApplyDecoratorsToElementDefinition(_proto_, _e_, _instanceMetadataList_, _instanceExtraInitializers_, *false*)).
+            1. If _result_ is an abrupt completion, then
+              1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+              1. Return ? _result_.
+        1. Let _protoMetaResult_ be Completion(ConvertDecoratorMetadataListToObject(_proto_, _instanceMetadataList_, _parentInstanceMetadata_)).
+        1. If _protoMetaResult_ is an abrupt completion, then
+          1. Set the running execution context's PrivateEnvironment to _outerPrivateEnvironment_.
+          1. Return ? _protoMetaResult_.
+        1. Set _F_.[[Elements]] to _instanceElements_.
+        1. Set _F_.[[Initializers]] to _instanceExtraInitializers_.
         1. For each PrivateElement _method_ of _staticPrivateMethods_, do
           1. Perform ! PrivateMethodOrAccessorAdd(_F_, _method_).
```

Comment:
You removed the definition of `PrivateMethodOrAccessorAdd`, so shouldn't be calling it.

Reply:
It doesn't seem like the spec initializes static methods anymore. Was `InitializePrivateMethods` intended to replace this as well? If so, you're missing a call to it.

Reply:
Apart from `PrivateMethodOrAccessorAdd` being missing, should static private methods be defined on the value of `F` before or after `ApplyDecoratorsToClassDefinition`? The current spec text seems to imply old `F`, but I think new `F` is the intended behavior and would make more sense.

Reply:
Hmm, that is tricky, the idea was that class decorators would run after method definition but before field definition. I think private methods work more like readonly fields though in a lot of ways, and it definitely doesn't make sense to have them be defined on a different object than fields, so I think you're correct, they should be defined after decorators have run.

Reply:
Yeah, it's a tricky area and I haven't thought it through, but it seems clear to me that you want to define private static methods on the new `F`, which is also the _classBinding_, otherwise you can't access the method in the usual way (`C.#staticMethod()`).